src/Chainweb/Chainweb/MinerResources.hs

```diff
@@ -68,7 +70,8 @@ data MiningCoordination logger cas = MiningCoordination
     , _coordCutDb :: !(CutDb cas)
     , _coordState :: !(TVar MiningState)
     , _coordLimit :: !Int
-    , _coord503s :: IORef Int }
+    , _coord503s :: IORef Int
+    , _coordChains :: TVar [ChainId] }
```

Comment:
The goal here is to avoid mining twice on the same parent. Instead of making chain selection completely deterministic, we should just do a minimal solution.

I propose:

```suggestion
    , _coordChains :: !(TVar (Vector (ChainId, BlockHash))) }
```

Each time we mine a block, we update the Vector with the block hash. When we look for new work we consider a chain as blocked if the block has for the chain in the cut is equal to the respective hash in `_coordChains`. So, we just extend the notion of "blocked" chain which already exists in the code.

The selection among the non-blocked chains remains the same `radomChainId`.

That has the advantage of less congestion: we update the `TVar` only when we mine a block (which is rare) and read only when we assign work. It also works with more than 10 miners. 

A deadlock can only happen when all nodes in the network get stuck at the same time with validation failure of block that they mind themselves, which is unlikely (and would destroy the network in any case).

Reply:
> Instead of making chain selection completely deterministic

It isn't truly deterministic, since chains would often be skipped due to blockage / cuts coming in from other nodes.

> Each time we mine a block, we update the Vector with the block hash. 

Mine a block, or accept a new cut? Mined blocks are not definitive.